Components version mapping ADR (#72)
opendatahub-io/architecture-decision-records: architecture-decision-records/operator/ODH-ADR-Operator-0007-components-version-mapping.md

> # Open Data Hub - Map component upstream versions to RHOAI releases

Reviewer: should it be ODH releases?

Author: yep, have changed this

|                |                                  |
| -------------- |----------------------------------|
| Date           | 2024-10-31                       |
| Scope          | Open Data Hub Operator           |

> | Status         | Approved                         |

Reviewer: already approved?

| Authors        | [Saravana Srinivasan](@sasriniv) |
| Supersedes     | N/A                              |
| Superseded by: | N/A                              |
| Tickets        |                                  |
| Other docs:    | none                             |

## What

> This document is intended to outline the design decisions made to map upstream versions of components that are supported by Data Science Cluster to RHOAI releases.

Reviewer: same here and below: if we are not only enable this for RHOAI, should ODH be used than RHOAI, same to all other ADR

## Why

Customers are expecting to know the list of upstream components and their versions they are using while using RHOAI. There were already several requests from the customers wanting to know the component versions during releases. 

## Goals

- To maintain a standard to capture the list of upstreams of the components along with their version and repository url that are supported by Data Science Cluster.
- Have it displayed in the ODH operator, precisely in Data Science Cluster's components status.

> - This information can then be read by the Dashboard UI via API and displayed in "About" modal of the UI.

Reviewer: i do not think this need to be as a goal.

## How

- Expecting components to create and maintain an "releases.yaml" file in their repositories particularly in manifest or config folders. The yaml can contain release details of the upstreams in format specified below.
```
releases:
  - displayname:
    version:
    repositoryurl:
  - displayname:
    version:
    repositoryurl:
```
- Develop code logics on the ODH operator to read through the releases.yaml file present in the manifest or config folder.
- Update the component status section in Data Science Cluster with the information read from the yaml file

> - Create UX designs to display the list of information shared by the operator.

Reviewer: same as here, this related to a specific feature in Dashboard.

Author: yep, removed

- Develop UI components implementing the UX designs to display the release information in the "About" modal.

## Open Questions

- **Updating version details:** The component team is responsible for creating and maintaining the "releases.yaml" file containing details of the upstreams and also, to promptly update the same when there are new additions to the upstreams.

## Alternatives

- Initially, discussion happened to maintain this version information in an .env file. But, with further discussions and understandings, certain components have got multiple upstreams with them. Maintaining and fetching multiple information from an env would be cumbersome, this pushed us to choose yaml file where we can group details of each of the upstream.

## Stakeholder Impacts

| Group                         | Key Contacts     | Date       | Impacted? |
| ----------------------------- | ---------------- | ---------- | --------- |

> | ODH Platform Team             | Saravana Srinivasan   |   | ? |

Reviewer 2: I think for key contacts here add @lburgazzoli @lphiri

Reviewer: should component team leads also set to be stakeholder?

Author: added component team leads

# Open Data Hub - Map component upstream versions to ODH releases

|                |                                  |
| -------------- |----------------------------------|
| Date           | 2024-10-31                       |
| Scope          | Open Data Hub Operator           |
| Status         | Draft                         |
| Authors        | [Saravana Srinivasan](@sasriniv) |
| Supersedes     | N/A                              |
| Superseded by: | N/A                              |
| Tickets        |                                  |
| Other docs:    | none                             |

## What

This document is intended to outline the design decisions made to map upstream versions of components that are supported by Data Science Cluster to ODH releases.

## Why

> Customers are expecting to know the list of upstream components and their versions they are using while using the product. There were already several requests from the customers wanting to know the component versions during releases. 

Reviewer 3: ```suggestion
Users are expecting to know the list of upstream components and their versions they are using while using the product. There were already several requests from users wanting to know the component versions during releases. 
```

^I think this might be a more upstream-friendly term?

## Goals

- To maintain a standard to capture the list of upstreams of the components along with their version and repository url that are supported by Data Science Cluster.

> - Have it displayed in the ODH operator, precisely in Data Science Cluster's components status.

Reviewer 3: ```suggestion
- Have it displayed in the Data Science Cluster's components status.
```

## How

> - Expecting components to create and maintain an "releases.yaml" file in their repositories particularly in manifest or config folders. The yaml can contain release details of the upstreams in format specified below.

Reviewer 3: The proposed file structure has a top-level key (currently `releases`) to hold the list. I like this approach as it allows for further information to be included in future (where an immediate list at the top level wouldn't have been as flexible). If any information was to be added in future though, would the filename `releases.yaml` still always make sense? 

I'd like to hear other opinions on this. I wonder if something more generic like `odh_metadata.yaml` would be better?

Author: +1 for a generic filename

Author: have changed the file name to odh_metadata.yaml

> - Expecting components to create and maintain an "odh_metadata.yaml" file in their repositories particularly in manifest or config folders. The yaml can contain release details of the upstreams in format specified below.

Reviewer 4: I would try to make the location of the file as stable as possible so ideally it should always be in a specific folder to reduce variance among how the odh operator handles the various components.

Author: In my understanding, components have just got either config or manifest folder, not both. for example kuberay has config [folder](https://github.com/opendatahub-io/kuberay/tree/dev/config)  in them and no manifests. and [getallManifests](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/get_all_manifests.sh) reads/downloads from the respective folders.

Reviewer 4: Yep but as I read it, I can have both, a new component can even use a different folder for manifests. I would suggest to rephrase this sentence to stress that the file should be added to the same location as the manifests retrieved by the odh-operator (independently on the folder name), ideally on the root.

Author: @lburgazzoli yea this makes more sense. I just rephrased it. Thanks :)

> - Expecting components team to create and maintain an "component_metadata.yaml" file in their repositories, in the same location as the manifests are being retrieved by the odh-operator, precisely in the root level. The yaml can contain release details of the upstreams in format specified below.

Reviewer 3: ```suggestion
- Component teams are expected to create and maintain a "component_metadata.yaml" file in their repositories, in the root of the manifests directory, which is retrieved by the ODH operator at build time. The yaml can contain release details of the upstreams in the format specified below.
```

Author: @grdryn Here some components keep their manifests in "config" directory, so thinking to rewrite a bit, like this "in the root of the directory from where the manifests are retrieved by the ODH operator at build time." wdyt?

Reviewer 3: > "in the root of the directory from where the manifests are retrieved by the ODH operator at build time."

:+1: 

/me wonders if we've got a common term for this location? :thinking: I guess from the get_all_manifests.sh it's "source-folder", but that might just be an internal implementation detail https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/get_all_manifests.sh#L7

```
releases:
  - name:
    version:
    repositoryurl:
  - name:
    version:
    repositoryurl:
```

> - Develop code logics on the ODH operator to read through the component_metadata.yaml file present in the same location as the manifests are being retrieved by the odh-operator, precisely in the root level.

Reviewer 3: ```suggestion
- Develop code logic on the ODH operator to read through the component_metadata.yaml file.
```

^Don't need to elaborate on the location of the file again, since that was just explained on line 29

- Update the component status section in Data Science Cluster with the information read from the yaml file

## Open Questions

> - **Updating version details:** The component team is responsible for creating and maintaining the "component_metadata.yaml" file containing details of the upstreams and also, to promptly update the same when there are new additions to the upstreams.

Reviewer 3: This doesn't appear to be a question, so I'm not sure why it's in the "Open Questions" section?

## Alternatives

> - Initially, discussion happened to maintain this version information in an .env file. But, with further discussions and understandings, certain components have got multiple upstreams with them. Maintaining and fetching multiple information from an env would be cumbersome, this pushed us to choose yaml file where we can group details of each of the upstream.

Reviewer 3: ```suggestion
- Initially, it was proposed to maintain this version information in an .env file. But, with further discussions and understanding, certain components have multiple upstreams with them. Maintaining and fetching information about multiple upstream releases from an env file would be cumbersome, this pushed us to choose a yaml file where we can group details of each of the upstreams.
```

| ------------------------- | --------------------------------------------------------------- | ---------- | --------- |
| ODH Platform Team         | @lburgazzoli @lphiri                                            |            | y         |
| Model Serving             | [Daniel Zonca](@danielezonca), [Edgar Hernández](@israel-hdez)  |            | y         |
| ODH Dashboard Team        | @andrewballantyne                                               |            | y         |
| IDE Team                  | @harshad16                                                      |            | y         |
| DS Pipelines Team         | @HumairAK @gmfrasca                                             |            | y         |
| Serving Team              | @Jooho                                                          |            | y         |
| TrustyAI Team             | @RobGeada                                                       |            | y         |
| Distributed Workloads     | @dimakis @anishasthana                                          |            | y         |

## References

> RHOAISTRAT-327 Refinement Document

Reviewer 3: Should this link to that document?

# Open Data Hub - Map component upstream versions to ODH releases

|                |                                  |
| -------------- |----------------------------------|
| Date           | 2024-10-31                       |
| Scope          | Open Data Hub Operator           |
| Status         | Draft                         |
| Authors        | [Saravana Srinivasan](@sasriniv) |
| Supersedes     | N/A                              |
| Superseded by: | N/A                              |
| Tickets        |                                  |
| Other docs:    | none                             |

## What

This document is intended to outline the design decisions made to map upstream versions of components that are supported by Data Science Cluster to ODH releases.

## Why

> Users are expecting to know the list of upstream components and their versions they are using while using the product. There were already several requests from users wanting to know the component versions during releases. 

Reviewer 5: "While they are using the product" is incorrect for this ADR. We are telling them what we shipped, not what is running on the cluster

Author: @Xaenalt yep, makes sense. Corrected it 👍

> Users are expecting to know the list of upstream components and their versions they are using while using the product. There were already several requests from the customers wanting to know the component versions during releases. 

Reviewer 3: ```suggestion
Users are expecting to know the list of upstream components and their versions they are using while using the product. There were already several requests from users wanting to know the component versions during releases. 
```

You got one, but not the other :slightly_smiling_face:

Author: oopss, yea just changed it. Thanks :)

## Goals

- To maintain a standard to capture the list of upstreams of the components along with their version and repository url that are supported by Data Science Cluster.
- Have it displayed in the Data Science Cluster's components status.

## How

- Component teams are expected to create and maintain a "component_metadata.yaml" file in their repositories, in the root of the directory from where the manifests are retrieved by the ODH operator at build time. The yaml can contain release details of the upstreams in the format specified below.
```
releases:
  - name:
    version:
    repositoryurl:
  - name:
    version:

>     repositoryurl:

Reviewer 2: Hi @harshad16 , just including the yaml file in the folder should be enough

>     repositoryurl:

Reviewer 6: ack, thanks @VaishnaviHire and @Sara4994 for clearing all my doubts.

N/A

## Responsibility

- **Updating version details:** The component teams are responsible for creating and maintaining the "component_metadata.yaml" file containing details of the upstreams and also, to promptly update the same when there are new additions to the upstreams.

## Alternatives

- Initially, it was proposed to maintain this version information in an .env file. But, with further discussions and understanding, certain components have multiple upstreams with them. Maintaining and fetching information about multiple upstream releases from an env file would be cumbersome, this pushed us to choose a yaml file where we can group details of each of the upstreams.

## Stakeholder Impacts

| Group                     | Key Contacts                                                    | Date       | Impacted? |
| ------------------------- | --------------------------------------------------------------- | ---------- | --------- |
| ODH Platform Team         | @lburgazzoli @lphiri                                            |            | y         |
| Model Serving             | [Daniel Zonca](@danielezonca), [Edgar Hernández](@israel-hdez)  |            | y         |

> | ODH Dashboard Team        | @andrewballantyne                                               |            | y         |

Reviewer 7: This LGTM from a design perspective. 👍 

cc @ederign

> | IDE Team                  | @harshad16                                                      |            | y         |

Reviewer 6: lgtm for design :+1:
